nomad/core_sched_test.go

```diff
@@ -439,26 +441,68 @@ func TestCoreScheduler_EvalGC_Batch_OldVersion(t *testing.T) {
 	alloc2.DesiredStatus = structs.AllocDesiredStatusRun
 	alloc2.ClientStatus = structs.AllocClientStatusLost
 
-	// Insert alloc with older job modifyindex
-	alloc3 := mock.Alloc()
-	job2 := job.Copy()
+	err = store.UpsertAllocs(structs.MsgTypeTestSetup, jobModifyIdx+2, []*structs.Allocation{alloc, alloc2})
+	if err != nil {
+		t.Fatalf("err: %v", err)
+	}
 
-	alloc3.Job = job2
-	alloc3.JobID = job2.ID
+	// Insert allocs with indexes older than job.ModifyIndex. Two cases:
+	//  1. Terminal state
+	//  2. Non-terminal state
+	alloc3 := mock.Alloc()
+	alloc3.Job = job
+	alloc3.JobID = job.ID
 	alloc3.EvalID = eval.ID
-	job2.CreateIndex = 500
 	alloc3.DesiredStatus = structs.AllocDesiredStatusRun
 	alloc3.ClientStatus = structs.AllocClientStatusLost
 
-	err = store.UpsertAllocs(structs.MsgTypeTestSetup, 1002, []*structs.Allocation{alloc, alloc2, alloc3})
+	alloc4 := mock.Alloc()
+	alloc4.Job = job
+	alloc4.JobID = job.ID
+	alloc4.EvalID = eval.ID
+	alloc4.DesiredStatus = structs.AllocDesiredStatusRun
+	alloc4.ClientStatus = structs.AllocClientStatusRunning
+
+	err = store.UpsertAllocs(structs.MsgTypeTestSetup, jobModifyIdx-1, []*structs.Allocation{alloc3, alloc4})
 	if err != nil {
 		t.Fatalf("err: %v", err)
 	}
 
-	// Update the time tables to make this work
-	tt := s1.fsm.TimeTable()
-	tt.Witness(2000, time.Now().UTC().Add(-1*s1.config.EvalGCThreshold))
+	// A little helper for assertions
+	assertCorrectEvalAlloc := func(
+		ws memdb.WatchSet,
+		eval *structs.Evaluation,
+		allocsShouldExist []*structs.Allocation,
+		allocsShouldNotExist []*structs.Allocation,
+	) {
+		out, err := store.EvalByID(ws, eval.ID)
+		if err != nil {
+			t.Fatalf("err: %v", err)
+		}
+		if out == nil {
+			t.Fatalf("bad: %v", out)
+		}
```

Comment:
We're migrating over to https://github.com/shoenig/test as we touch old tests and write new ones. For all the new assertions, let's use that.

```suggestion
		must.NoError(t, err)
		must.NotNil(t, out)
```